pkg/controller/mpi_job_controller.go

```diff
@@ -656,13 +656,13 @@ func (c *MPIJobController) syncHandler(key string) error {
 				return err
 			}
 		}
-		if mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
+		// The Intel and MPICH implementations require workers to communicate with the
+		// launcher through its hostname. For that, we create a Service which
+		// has the same name as the launcher's hostname.
+		if (mpiJob.Spec.RunLauncherAsWorker != nil && *mpiJob.Spec.RunLauncherAsWorker) ||
+			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationIntel ||
 			mpiJob.Spec.MPIImplementation == kubeflow.MPIImplementationMPICH {
```

Comment:
Uhm, I'm wondering if we can make this condition another function like `workersCanHaveDedicatedService` since we're using the same condition in some places.